File: Dokie/ViewControllers/Document/DocumentViewModel.m
//
//  DocumentViewModel.m
//  Dokie
//
//  Created by Ashamaz Shidov on 28/12/2019.
//  Copyright © 2019 Ashamaz Shidov. All rights reserved.
//

#import "DocumentViewModel.h"
#import "DocumentFilesViewModel.h"
#import "DocumentTextViewModel.h"
#import "CoreDataInclude.h"
#import "NavigationRouter.h"
#import "CoreDataManager.h"

#import <UIKit/UIKit.h>

@interface DocumentViewModel()

@property (nonatomic, strong) CDDocument *document;
@property (nonatomic, strong) NSString *documentName;
@property (nonatomic, strong) DocumentFilesViewModel *filesViewModel;
@property (nonatomic, strong) DocumentTextViewModel *textViewModel;

@property (nonatomic, strong, readonly) NavigationRouter *navigationRouter;
@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;

@end

@implementation DocumentViewModel

- (instancetype)initWithDocument:(CDDocument *)document {
    self = [super init];
    if (self) {
        self.document = document;
        [self setup];
    }
    return self;
}

- (void)setup {
    self.filesViewModel = [[DocumentFilesViewModel alloc] initWithDocument:self.document];
    self.textViewModel = [[DocumentTextViewModel alloc] initWithDocument:self.document];
    @weakify(self);
    [[[self.coreDataManager documentName:self.document] deliverOnMainThread] subscribeNext:^(NSString * _Nullable x) {
        @strongify(self);
        self.documentName = x;
    }];
}

- (void)addButtonPressed {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleActionSheet];
    
    @weakify(self);
    UIAlertAction *addText = [UIAlertAction actionWithTitle:NSLocalizedString(@"Add text", @"Add text button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self.textViewModel addContent];
    }];
    [alert addAction:addText];
    UIAlertAction *addFile = [UIAlertAction actionWithTitle:NSLocalizedString(@"Add file", @"Add file button text in alert view") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self);
        [self.filesViewModel addFile];
    }];
    [alert addAction:addFile];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button text in alert view") style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:cancel];
    [self.navigationRouter showAlert:alert];

}

- (void)shareText:(NSArray<NSString *> *)texts images:(NSArray<CDContent *> *)images {
    if (texts.count == 0 && images.count == 0) return;
    
    NSMutableString *string = [NSMutableString string];
    for (NSString *text in texts) {
        [string appendFormat:@"%@\n", text];
    }

    if (images.count > 0) {
        NSString *loadingGuid = [self.navigationRouter showLoading];
        NSMutableArray *signals = [NSMutableArray array];
        for (CDContent *content in images) {
            [signals addObject:[self.coreDataManager imageFromContent:content]];
        }
        
        @weakify(self);
        [[[RACSignal combineLatest:signals] deliverOnMainThread] subscribeNext:^(RACTuple * _Nullable x) {
            @strongify(self);
            NSMutableArray *imagesArray = [NSMutableArray array];
            for (UIImage *image in x) {
                [imagesArray addObject:image];
            }
            
            if (string.length > 0) {
                [imagesArray addObject:[string copy]];
            }
            [self.navigationRouter shareItems:imagesArray];
            [self.navigationRouter hideLoading:loadingGuid];
        }];
    }
    else {
        [self.navigationRouter shareItems:@[[string copy]]];
    }
}

#pragma mark - Get Set

- (NavigationRouter *)navigationRouter {
    return [NavigationRouter shared];
}

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

@end

File: Dokie/ViewControllers/Document/DocumentViewModel.swift
//
//  DocumentViewModel.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 25.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class DocumentViewModel: NSObject {
    
    let filesViewModel: DocumentFilesViewModel
    let textViewModel: DocumentTextViewModel
    let tagsViewModel: TagsScrollViewModel
    let documentName = MutableProperty<String>("")
    let forceSegmentedSectionSelect = MutableProperty<Int>(0)

    private let document: CDDocument
    private let navigationRouter = NavigationRouter.shared
    private let coreDataManager = CoreDataManager.shared
    
    @available(*, unavailable)
    override init() {
        fatalError("Not implemented")
    }
    
    init(document: CDDocument) {
        self.document = document
        self.filesViewModel = DocumentFilesViewModel(document: document)
        self.textViewModel = DocumentTextViewModel(document: document)
        self.tagsViewModel = TagsScrollViewModel(document: document)
        super.init()
        setup()
    }
    
    private func setup() {
        coreDataManager
            .name(of: document)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case let .success(name):
                    self.documentName.value = name
                default:
                    self.documentName.value = ""
                }
        }
    }
    
    func addButtonPressed() {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let addText = UIAlertAction(title: String.localized("Add text"), style: .default) { [unowned self] _ in
            self.forceSegmentedSectionSelect.value = 0
            self.textViewModel.addContent()
        }
        let addFile = UIAlertAction(title: String.localized("Add file"), style: .default) { [unowned self] _ in
            self.forceSegmentedSectionSelect.value = 1
            self.filesViewModel.addFile()
        }
        let cancel = UIAlertAction(title: String.localized("Cancel"), style: .cancel)
        
        alert.addAction(addText)
        alert.addAction(addFile)
        alert.addAction(cancel)
        navigationRouter.showAlert(alert)
    }
    
    func share(texts: [String], images: [CDContent]) {
        guard texts.count > 0 || images.count > 0 else { return }
        
        var string = ""
        for text in texts {
            string = string + "\(text)\n"
        }
        
        if images.count > 0 {
            let loadingGiud = navigationRouter.showLoading()
            var signals = [SignalProducer<UIImage?, Never>]()
            for content in images {
                signals.append(coreDataManager.imageFrom(content: content).flatMapError({ _ -> SignalProducer<UIImage?, Never> in
                    SignalProducer<UIImage?, Never>.empty
                }))
            }
            
            SignalProducer
                .combineLatest(signals)
                .take(during: reactive.lifetime)
                .observe(on: UIScheduler())
                .startWithValues { [unowned self] images in
                    var imagesArray = [Any]()
                    for image in images {
                        if let image = image {
                            imagesArray.append(image)
                        }
                    }
                    if !string.isEmpty {
                        imagesArray.append(string)
                    }
                    
                    self.navigationRouter.share(items: imagesArray)
                    self.navigationRouter.hideLoading(loadingGiud)
            }
        }
        else if !string.isEmpty {
            navigationRouter.share(items: [string])
        }
    }
    
}
